lib/writable.js

```diff
@@ -1,69 +1,225 @@
+
+    if (chunk.length < remainingSize) {
+      this.writeChunk(chunk, encoding);
+    } else {
+      const currentBufferChunk = chunk.slice(0, this.remainingSize);
+      this.writeChunk(currentBufferChunk);
+      this.queue.push(this.buffer);
+      this.buffer = Buffer.alloc(this.buffer.length);
+      this.offset = 0;
+
+      let nextBufferChunk = chunk.slice(remainingSize);
+
+      if (nextBufferChunk < this.buffer.length) {
+        this.writeChunk(nextBufferChunk, encoding);
+      } else {
+        if (typeof nextBufferChunk === 'string') {
+          nextBufferChunk = Buffer.from(nextBufferChunk, encoding);
+        }
+        this.queue.push(nextBufferChunk);
+        this.writableLength += nextBufferChunk.length;
+      }
+    }
+  }
+
+  getBuffer() {
+    const bufferData = this.buffer.slice(0, this.offset);
+    this.queue.push(bufferData);
+    const ret = this.queue;
+    this.queue = [];
+    this.offset = 0;
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (!this.corked && this.writableLength) {
+      this.writing = true;
+      this.doWrite(this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    this.defaultEncoding = encoding;
+  }
+
+  pipe() {
+    this.emit('error', new Error('[ERR_STREAM_CANNOT_PIPE]'));
+  }
+
+  validChunk(chunk, cb) {
+    let error;
+    if (chunk === null) {
+      error = new Error('[ERR_STREAM_NULL_VALUES]');
+    } else if (typeof chunk !== 'string' && !Buffer.isBuffer(chunk)) {
+      error = new Error('[ERR_INVALID_ARG_TYPE]');
+    }
+
+    if (error) {
+      this.emit('error', error);
+      process.nextTick(cb, error);
+      return false;
+    }
+    return true;
+  }
+
+  write(chunk, encoding, cb) {
+    if (typeof encoding === 'function') {
+      cb = encoding;
+      encoding = null;
+    }
+
+    if (this.ended) {
+      const err = new Error('[ERR_STREAM_WRITE_AFTER_END]');
+      this.emit('error', err);
+      process.nextTick(cb, err);
+    }
+
+    if (this.destroyed) {
+      const err = new Error('[ERR_STREAM_DESTROYED]');
+      this.emit('error', err);
+      process.nextTick(cb, err);
+    }
+
+    if (!encoding) {
+      encoding = this.defaultEncoding;
+    }
+
+    if (typeof cb !== 'function') {
+      cb = () => {};
```

Comment:
Use `common.safeCallback` or `common.once`. But `safeCallback` takes array `args` so we may need `safeFunction`: ```js
const safeFunction = fn => typeof fn === 'function' ? fn : emptiness;
```

Reply:
Do you want to add a safeFunction to a stream library or to common?

Reply:
yes

Reply:
Why does this assignment is even needed? Can't you just add 
`if (cb) ...` on line [117](https://github.com/metarhia/metastreams/pull/9/files#diff-0c3f400a496e158d7c5c7899e9fc54f9R117) and later on in this function?
Also, IMO having empty callback here is harmful as 
* it will grow the `callbacks` array needlessly
* you will have it scheduled on nextTick in case of error `process.nextTick(cb, error);`